### ISSUE TITLE
BUG: pd.DataFrame.to_hdf does not support multiindex frames with `string[pyarrow_numpy]`

### ISSUE THREAD

**Issue by nvasilevv** [pandas-dev/pandas#55088]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
data = pd.DataFrame(
    np.random.randint(0, 50, size=(15, 4)), columns=list('ABCD')
).set_index(['A', 'B'])

# This action will be successful as it works on numpy objects
data.to_hdf('test.hdf', key='test', format='table')

# If we force the usage of pyarrow string, the same will fail

pd.options.future.infer_string = True

data_arrow = pd.DataFrame(
    np.random.randint(0, 50, size=(15, 4)), columns=list('ABCD')
).set_index(['A', 'B'])

data_arrow.to_hdf('test_arrow.hdf', key='test', format='table') # This will raise ValueError("cannot have non-object label DataIndexableCol")
```

### Issue Description

Looks like using `string[pyarrow_numpy]` as dtype and `to_hdf` with a multiindex frame throws a value error. The error comes from `pytables.py` and more specifically the check below.

https://github.com/pandas-dev/pandas/blob/723feb984e6516e3e1798d3c4440c844b12ea18f/pandas/io/pytables.py#L2647-L2650

The `is_object_dtype` always checks against `np.object_`.

https://github.com/pandas-dev/pandas/blob/723feb984e6516e3e1798d3c4440c844b12ea18f/pandas/core/dtypes/common.py#L137-L165

### Expected Behavior

I would expect that `is_object_dtype` allows for `string[pyrrow_numpy]` (if and only if downstream behavior is supported by that dtype). If that action is not supported with `string[pyarrow_numpy]` I would expected a more informative error.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : 723feb984e6516e3e1798d3c4440c844b12ea18f
python              : 3.9.13.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.19044
machine             : AMD64
processor           : Intel64 Family 6 Model 165 Stepping 2, GenuineIntel  
byteorder           : little
LC_ALL              : None
LANG                : en_US.UTF-8
LOCALE              : English_United States.1252

pandas              : 2.2.0dev0+208.g723feb984e
numpy               : 1.22.4
pytz                : 2022.7.1
dateutil            : 2.8.2
setuptools          : 68.0.0
pip                 : 23.2.1
Cython              : 0.29.36
pytest              : 6.2.5
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2
IPython             : None
pandas_datareader   : None
bs4                 : None
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : None
gcsfs               : None
matplotlib          : None
numba               : 0.56.4
numexpr             : 2.8.5
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : 10.0.1
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.10.1
sqlalchemy          : None
tables              : 3.7.0
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : None
pyqt5               : None

</details>

**Comment by phofl** [comment]
Thx for the report. Your assumption is correct, this should work

**Comment by nvasilevv** [comment]
I think that I could try to propose a PR fixing this. It would be my first one in this project, so if you think that the topic is too hard to approach, I am fine with waiting for the solution.

**Comment by phofl** [comment]
I'd appreciate the help if you have time. The first thing I would do is figuring out where "pyarrow[string]" is treated differently than the new dtype.

**Comment by nvasilevv** [comment]
After undertaking a short investigation, I believe that the different treatment in that case comes from here

https://github.com/pandas-dev/pandas/blob/705d4312cf1d94ef2497bcd8091e0eabd1085f4a/pandas/_libs/lib.pyx#L2700C12-L2706

Essentially, this part always uses `pyarrow_numpy` as a string type if `pd.options.future.infer_string = True`. However, if that is not set and instead `pd.options.mode.string_storage = 'pyarrow'` nothing changes.

I see that this part was changed here - https://github.com/pandas-dev/pandas/commit/1ecfa6749d07a08d501d56b98ae60d483cf1078f and was released in v2.1.0. Previously, the array was not casted at this point.
